teal/packer.tl

```diff
@@ -0,0 +1,130 @@
+local M = {}
+
+local function apply_config(plugin: Plugin)
+  if plugin.config and plugin.loaded then
+    local c = plugin.config
+    if c is function() then
+      c()
+    else
+      loadstring(c, plugin.name..'.config()')()
+    end
+  end
+end
+
+local function loader(lplugins: {Plugin})
+  for _, plugin in ipairs(lplugins) do
+    if not plugin.loaded then
+      -- Set the plugin as loaded before config is run in case something in the
+      -- config tries to load this same plugin again
+      plugin.loaded = true
+      vim.cmd.packadd(plugin.name)
+      apply_config(plugin)
+    end
+  end
+end
+
+-- Completion user plugins
+-- Intended to provide completion for PackerUpdate/Sync/Install command
+local function plugin_complete(lead: string, _:string): {string}
+  local completion_list = vim.tbl_filter(function(name: string): boolean
+    return vim.startswith(name, lead)
+  end, vim.tbl_keys(_G.packer_plugins))
+  table.sort(completion_list)
+  return completion_list
+end
+
+local function load_plugin_configs()
+  local Handlers = require('packer.handlers')
+
+  local cond_plugins: {Handlers.Type:{string:Plugin}} = {
+    cmd   = {},
+    keys  = {},
+    ft    = {},
+    event = {},
+  }
+
+  local uncond_plugins = {}
+
+  local conds = {'cmd', 'keys', 'ft', 'event'} as {Handlers.Type}
+
+  for name, plugin in pairs(_G.packer_plugins) do
+    local has_cond = false
+    for _, cond in ipairs(conds) do
+      if (plugin as {string:any})[cond] then
+        has_cond = true
+        cond_plugins[cond][name] = plugin
+        break
+      end
+    end
+    if not has_cond then
+      uncond_plugins[name] = plugin
+    end
+  end
+
+  for _, plugin in pairs(uncond_plugins) do
+    apply_config(plugin)
+  end
+
+  for _, cond in ipairs(conds) do
+    if next(cond_plugins[cond]) then
+      Handlers[cond](cond_plugins[cond], loader)
+    end
+  end
+end
+
+local function make_commands()
+  local actions = setmetatable({}, {
+    __index = function(_, k: string): function
+      return function(...:any): any...
+        return (require'packer.actions' as {string:function})[k](...)
+      end
+    end
+  }) as {string:function}
+
+  for _, cmd in ipairs {
+    { 'PackerInstall', '*', actions.install , plugin_complete},
+    { 'PackerUpdate' , '*', actions.update  , plugin_complete},
+    { 'PackerSync'   , '*', actions.sync                     },
+    { 'PackerClean'  , '*', actions.clean                    },
+    { 'PackerStatus' , '*', actions.status                   },
+  } do
+    vim.api.nvim_create_user_command(cmd[1], function(args: vim.api.UserCmdParams)
+      cmd[3](unpack(args.fargs))
+    end, { nargs = cmd[2], complete = cmd[4] })
+  end
+end
+
+local record SSpec
+  {Plugin}
+
+  config: Config
+end
+
+-- Convenience function for simple setup
+-- spec can be a table with a table of plugin specifications as its first
+-- element, config overrides as another element.
+function M.startup(spec: SSpec)
```

Comment:
May I recommend breaking with the vim-plug inherited API (which probably most people looking to packer won't even know about anymore) and instead doing something like
* `packer.setup(spec, config)` (by now the common name for a plugin's entry point)
* `packer.add(spec)` (makes it clearer that plugins are added to the managed list)

In particular, this split of `spec` and `config` prevents the global `config` from needing to be merged (or overwritten by accident).

Reply:
> Not necessarily in this PR, but we should also clarify that you can still get the finer-grained control of the old init() configuration approach by passing only config overrides and no plugin specs to startup, then adding plugin specs with a (currently missing) use function/similar.

This wasn't the intention. I was working towards the user required to run `startup` just once, with no possibility of adding plugins later. However, the way I've written this will allow this (accidentally), so I guess we can support that.

I like @clason suggestion, let's try that.

Reply:
I also feel moderately strongly that users should be able to add plugins outside of the initial call to `startup`, since I have encountered a good number of `packer` users who do so for greater flexibility in controlling the managed set (specifically thinking of "layering" approaches in reusable configs here). There's an easy way around that if we require users to rewrite such configs to build their own table of plugins and pass it to us once, but it doesn't add much complexity (imo) on our end to support this.

I also like @clason's suggested signatures and naming. However, I want to note that we also need to expose a "run the loaders" function for anyone using the pattern, i.e.:

```lua
packer.setup({}, config)
packer.add(...)
packer.add(...)
packer.load()
```

Reply:
Would `packer.load()` be mandatory if you don't use `packer.add()`?

Reply:
Given the way this is written, I'm not sure we need a `load`.